WebXROpaqueFramebuffer queries state from GPUP

kkinnunen-apple opened the pull request:
#### 05bd97a27164d3373bd61101c2f995d292bc36f0
<pre>
WebXROpaqueFramebuffer queries state from GPUP
<a href="https://bugs.webkit.org/show_bug.cgi?id=264248">https://bugs.webkit.org/show_bug.cgi?id=264248</a>
<a href="https://rdar.apple.com/117992444">rdar://117992444</a>

Reviewed by Dan Glastonbury.

The context state is available in WebGLRenderingContextBase, and thus
it should not be queried from the underlying OpenGL context.

Add Scoped*Restore classes to encapsulate restoring the state from
the WebGL context. These will also be used when default framebuffer
implementation is moved from the underlying context to the webgl context
level.

Move all the scoped helpers to WebGLUtilities.h so they&apos;re available
to all the WebGL implementation files.

This is work towards supporting premultipliedAlpha=false compositing.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::endFrame):
(WebCore::WebXROpaqueFramebuffer::setupFramebuffer):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
(WebCore::WebGL2RenderingContext::drawRangeElements):
* Source/WebCore/html/canvas/WebGL2RenderingContext.h:
* Source/WebCore/html/canvas/WebGLDrawInstancedBaseVertexBaseInstance.cpp:
(WebCore::WebGLDrawInstancedBaseVertexBaseInstance::drawArraysInstancedBaseInstanceWEBGL):
(WebCore::WebGLDrawInstancedBaseVertexBaseInstance::drawElementsInstancedBaseVertexBaseInstanceWEBGL):
* Source/WebCore/html/canvas/WebGLMultiDraw.cpp:
(WebCore::WebGLMultiDraw::multiDrawArraysWEBGL):
(WebCore::WebGLMultiDraw::multiDrawArraysInstancedWEBGL):
(WebCore::WebGLMultiDraw::multiDrawElementsWEBGL):
(WebCore::WebGLMultiDraw::multiDrawElementsInstancedWEBGL):
* Source/WebCore/html/canvas/WebGLMultiDrawInstancedBaseVertexBaseInstance.cpp:
(WebCore::WebGLMultiDrawInstancedBaseVertexBaseInstance::multiDrawArraysInstancedBaseInstanceWEBGL):
(WebCore::WebGLMultiDrawInstancedBaseVertexBaseInstance::multiDrawElementsInstancedBaseVertexBaseInstanceWEBGL):
* Source/WebCore/html/canvas/WebGLRenderingContext.cpp:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::drawArrays):
(WebCore::WebGLRenderingContextBase::drawElements):
(WebCore::WebGLRenderingContextBase::drawArraysInstanced):
(WebCore::WebGLRenderingContextBase::drawElementsInstanced):
(WebCore::ScopedTightUnpackParameters::ScopedTightUnpackParameters): Deleted.
(WebCore::ScopedTightUnpackParameters::~ScopedTightUnpackParameters): Deleted.
(WebCore::ScopedTightUnpackParameters::set): Deleted.
(WebCore::ScopedDisableRasterizerDiscard::ScopedDisableRasterizerDiscard): Deleted.
(WebCore::ScopedDisableRasterizerDiscard::~ScopedDisableRasterizerDiscard): Deleted.
(WebCore::ScopedEnableBackbuffer::ScopedEnableBackbuffer): Deleted.
(WebCore::ScopedEnableBackbuffer::~ScopedEnableBackbuffer): Deleted.
(WebCore::ScopedDisableScissorTest::ScopedDisableScissorTest): Deleted.
(WebCore::ScopedDisableScissorTest::~ScopedDisableScissorTest): Deleted.
(WebCore::InspectorScopedShaderProgramHighlight::InspectorScopedShaderProgramHighlight): Deleted.
(WebCore::InspectorScopedShaderProgramHighlight::~InspectorScopedShaderProgramHighlight): Deleted.
(WebCore::InspectorScopedShaderProgramHighlight::showHighlight): Deleted.
(WebCore::InspectorScopedShaderProgramHighlight::hideHighlight): Deleted.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:

Canonical link: <a href="https://commits.webkit.org/270886@main">https://commits.webkit.org/270886@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ee2bd990a85fc465252453b2960eaf4aeea501be

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25185 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3726 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26441 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27300 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23109 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25454 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5439 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1162 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23348 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25428 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2752 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21737 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/2470 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2437 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22677 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27879 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22984 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23025 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28798 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2388 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/679 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26624 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3736 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6417 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2825 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2720 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebCore/html/canvas/WebGLDrawInstancedBaseVertexBaseInstance.cpp:
```diff
@@ -64,7 +64,7 @@ void WebGLDrawInstancedBaseVertexBaseInstance::drawArraysInstancedBaseInstanceWE
 
     {
-        InspectorScopedShaderProgramHighlight scopedHighlight(context, context.m_currentProgram.get());
+        ScopedInspectorShaderProgramHighlight scopedHighlight { context };
```
In future, I would prefer these extra changes to be in their own change, so that the title of the commit matches the bulk of the changes.

djg reviewed (approved):
:shipit:

kkinnunen-apple committed:
WebXROpaqueFramebuffer queries state from GPUP
https://bugs.webkit.org/show_bug.cgi?id=264248
rdar://117992444

Reviewed by Dan Glastonbury.

The context state is available in WebGLRenderingContextBase, and thus
it should not be queried from the underlying OpenGL context.

Add Scoped*Restore classes to encapsulate restoring the state from
the WebGL context. These will also be used when default framebuffer
implementation is moved from the underlying context to the webgl context
level.

Move all the scoped helpers to WebGLUtilities.h so they're available
to all the WebGL implementation files.

This is work towards supporting premultipliedAlpha=false compositing.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::endFrame):
(WebCore::WebXROpaqueFramebuffer::setupFramebuffer):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
(WebCore::WebGL2RenderingContext::drawRangeElements):
* Source/WebCore/html/canvas/WebGL2RenderingContext.h:
* Source/WebCore/html/canvas/WebGLDrawInstancedBaseVertexBaseInstance.cpp:
(WebCore::WebGLDrawInstancedBaseVertexBaseInstance::drawArraysInstancedBaseInstanceWEBGL):
(WebCore::WebGLDrawInstancedBaseVertexBaseInstance::drawElementsInstancedBaseVertexBaseInstanceWEBGL):
* Source/WebCore/html/canvas/WebGLMultiDraw.cpp:
(WebCore::WebGLMultiDraw::multiDrawArraysWEBGL):
(WebCore::WebGLMultiDraw::multiDrawArraysInstancedWEBGL):
(WebCore::WebGLMultiDraw::multiDrawElementsWEBGL):
(WebCore::WebGLMultiDraw::multiDrawElementsInstancedWEBGL):
* Source/WebCore/html/canvas/WebGLMultiDrawInstancedBaseVertexBaseInstance.cpp:
(WebCore::WebGLMultiDrawInstancedBaseVertexBaseInstance::multiDrawArraysInstancedBaseInstanceWEBGL):
(WebCore::WebGLMultiDrawInstancedBaseVertexBaseInstance::multiDrawElementsInstancedBaseVertexBaseInstanceWEBGL):
* Source/WebCore/html/canvas/WebGLRenderingContext.cpp:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::drawArrays):
(WebCore::WebGLRenderingContextBase::drawElements):
(WebCore::WebGLRenderingContextBase::drawArraysInstanced):
(WebCore::WebGLRenderingContextBase::drawElementsInstanced):
(WebCore::ScopedTightUnpackParameters::ScopedTightUnpackParameters): Deleted.
(WebCore::ScopedTightUnpackParameters::~ScopedTightUnpackParameters): Deleted.
(WebCore::ScopedTightUnpackParameters::set): Deleted.
(WebCore::ScopedDisableRasterizerDiscard::ScopedDisableRasterizerDiscard): Deleted.
(WebCore::ScopedDisableRasterizerDiscard::~ScopedDisableRasterizerDiscard): Deleted.
(WebCore::ScopedEnableBackbuffer::ScopedEnableBackbuffer): Deleted.
(WebCore::ScopedEnableBackbuffer::~ScopedEnableBackbuffer): Deleted.
(WebCore::ScopedDisableScissorTest::ScopedDisableScissorTest): Deleted.
(WebCore::ScopedDisableScissorTest::~ScopedDisableScissorTest): Deleted.
(WebCore::InspectorScopedShaderProgramHighlight::InspectorScopedShaderProgramHighlight): Deleted.
(WebCore::InspectorScopedShaderProgramHighlight::~InspectorScopedShaderProgramHighlight): Deleted.
(WebCore::InspectorScopedShaderProgramHighlight::showHighlight): Deleted.
(WebCore::InspectorScopedShaderProgramHighlight::hideHighlight): Deleted.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:

Canonical link: https://commits.webkit.org/270886@main
